File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
from collections import defaultdict
import heapq

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counter = defaultdict(int)
        for num in nums:
            counter[num] += 1

        pq = []
        for num,freq in counter.items():
            heapq.heappush(pq, (freq, num))

            if len(pq) > k:
                heapq.heappop(pq)
        
        ans = []
        while pq:
            _, num = heapq.heappop(pq)
            ans.append(num)
        return ans
    

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        var counter = new HashMap<Integer, Integer>();
        for (int num: nums) {
            counter.put(num, counter.getOrDefault(num, 0) + 1);
        }

        var buckets = new ArrayList<ArrayList<Integer>>();
        for (int i = 0; i < nums.length + 1; i++) {
            buckets.add(new ArrayList<Integer>());
        } 

        for (Map.Entry<Integer, Integer> entry: counter.entrySet()) {
            buckets.get(entry.getValue()).add(entry.getKey());
        }

        var topK = new int[k];
        var idx = 0;
        for (int i = nums.length; i >= 0; i--) {
            var bucket = buckets.get(i);
            for (int num: bucket) {
                topK[idx++] = num;
                if (idx == k) {
                    return topK;
                }
            }
        }

        return topK;
    }
}